dfget/core/uploader/peer_server.go

```diff
@@ -0,0 +1,455 @@
+/*
+ * Copyright The Dragonfly Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uploader
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/dragonflyoss/Dragonfly/dfget/config"
+	"github.com/dragonflyoss/Dragonfly/dfget/core/api"
+	"github.com/dragonflyoss/Dragonfly/dfget/core/helper"
+	"github.com/dragonflyoss/Dragonfly/dfget/errors"
+	"github.com/dragonflyoss/Dragonfly/dfget/util"
+	"github.com/dragonflyoss/Dragonfly/version"
+	"github.com/gorilla/mux"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+// newPeerServer return a new P2PServer.
+func newPeerServer(cfg *config.Config, port int) *peerServer {
+	s := &peerServer{
+		cfg:      cfg,
+		finished: make(chan struct{}),
+		host:     cfg.RV.LocalIP,
+		port:     port,
+		api:      api.NewSupernodeAPI(),
+	}
+
+	r := s.initRouter()
+	s.Server = &http.Server{
+		Addr:    net.JoinHostPort(s.host, strconv.Itoa(port)),
+		Handler: r,
+	}
+
+	return s
+}
+
+// ----------------------------------------------------------------------------
+// peerServer structure
+
+// peerServer offer file-block to other clients
+type peerServer struct {
```

Comment:
Could you add more description of some particular fields of struct peerServer?

Reply:
added